### PR TITLE
🐛 [amp-carousel 0.2] Stop propagation of `touchmove` event to viewer

### LINKED ISSUE #33433: [amp-carousel] nav swipe is overridden by top stories carousel
## What's the issue?
The horizontal swipe on our nav gets overridden by the scroll for the `Top Stories` carousel. The nav appears to need the browser's focus for the scroll to work, and this focus is taken by the top stories carousel. Historically, the only way we've been able to get horizontal swiping to work is by implementing our nav with `amp-carousel`. This implementation was working in the `Top Stories` carousel until recently.

## How do we reproduce the issue?

1. On a mobile device or in mobile view on your browser, google `nytimes coronavirus vaccine`.
2. Scroll down till you reach the top stories list.
3. Click into the first story. The `Coronavirus Outbreak` nav should appear on top of the story.
4. Try to swipe horizontally on the nav to view the other menu items. The horizontal swipe will instead take you to the next top story in the carousel.

## What browsers are affected?

Chrome, Safari, Firefox.

## Which AMP version is affected?

This was working until recently (a few weeks ago). We are using the latest version of `amp-carousel` -- https://cdn.ampproject.org/v0/amp-carousel-0.2.js

Slack conversation: https://amphtml.slack.com/archives/C9HPA6HGB/p1616452405064800

### PR DESCRIPTION
**Pull request by micajuine-ho**
Add a boolean signal, `shouldViewerCancelPropagation`, to `touchmove` events that occur on the `scrollContainer` of amp-carousel 0.2, when severed in a a viewer. 

Within `amp-viewer-integration`'s `touch-handler`, when an event that is being bubbled up, has the `shouldViewerCancelPropagation` signal, stop propagation so that the viewer doesn't get event.

Closes #33433

### CODE REVIEW

File: extensions/amp-carousel/0.2/amp-carousel.js [hunk 1]
```diff
@@ -632,6 +636,19 @@ class AmpCarousel extends AMP.BaseElement {
     );
   }
 
+  /**
+   * Attach singal onto touchmove event to stop propagation to viewer
+   * to be handled at the viewer integration layer in bubble phase.
+   */
+  handlePropagationInViewer_() {
+    if (!this.isViewerEmbeded_) {
```
Comment (kristoferbaxter) [inline extensions/amp-carousel/0.2/amp-carousel.js:644]: This is much cleaner than the version we discussed earlier (using capture phase of the event to pass this information). Nice job!

File: extensions/amp-carousel/0.2/test/test-type-slides.js [hunk 2]
```diff
@@ -529,5 +531,34 @@ describes.realWin(
         });
       });
     });
+
+    describe('event propogation', () => {
+      it('should add touchmove event if in viewer', async () => {
+        env.sandbox.stub(Services, 'viewerForDoc').returns({
+          isEmbedded: () => {
```
Comment (kristoferbaxter) [inline extensions/amp-carousel/0.2/test/test-type-slides.js:538]: Nit: `isEmbedded: () => true`

File: extensions/amp-carousel/0.2/test/test-type-slides.js [hunk 3]
```diff
@@ -529,5 +531,34 @@ describes.realWin(
...
+        const impl = await carousel.getImpl();
+
+        expect(listenSpy.lastCall.args[0]).to.equal(impl.scrollContainer_);
+        expect(listenSpy.lastCall.args[1]).to.equal('touchmove');
+        expect(listenSpy.args.length).to.equal(5);
+      });
+
+      it('should not add touchmove event if not in the viewer', async () => {
+        env.sandbox.stub(Services, 'viewerForDoc').returns({
+          isEmbedded: () => {
```
Comment (kristoferbaxter) [inline extensions/amp-carousel/0.2/test/test-type-slides.js:553]: Nit: `isEmbedded: () => false`

File: extensions/amp-viewer-integration/0.1/test/test-touch-handler.js [hunk 4]
```diff
@@ -114,6 +114,23 @@ describes.fakeWin('TouchHandler', {}, (env) => {
...
+    it('should not forward events marked with `shouldViewerCancelPropagation`', () => {
+      const event = fakeTouchEvent('touchmove');
+      event.stopImmediatePropagation = env.sandbox.spy();
+
+      touchHandler.handleEvent_(event);
+      expect(messages).to.have.length(1);
+      expect(messages[0].data.data.type).to.equal('touchmove');
+      expect(event.stopImmediatePropagation).to.not.be.called;
+
+      event.shouldViewerCancelPropagation = true;
```
Comment (kristoferbaxter) [inline extensions/amp-viewer-integration/0.1/test/test-touch-handler.js:126]: Do you think this test could be separated out a bit to ensure it's per touch event?

Case 1: Never stop propagation.
Case 2: Stop propagation once.
Case 3: Allow propagation once, stop propagation once.
Case 4: Stop propagation once, allow propagation once.

File: extensions/amp-viewer-integration/0.1/touch-handler.js [hunk 5]
```diff
@@ -138,6 +138,10 @@ export class TouchHandler {
    * @private
    */
   forwardEvent_(e) {
+    if (e && e.shouldViewerCancelPropagation) {
```
Comment (kristoferbaxter) [inline extensions/amp-viewer-integration/0.1/touch-handler.js:141]: I'd add a comment about conditions this occurs under so it's more obvious in six-months.

**Review by kristoferbaxter — APPROVED** [review]
Looks good to me. Left a few nits and items that can be addressed in a future PR.

**Comment by mspr1nt** [comment]
Just popping in here from the Slack channel to add some examples of this.  Searches are all being made straight from the Chrome Android browser. It appears though the issue might not occur when the first story in the Top Stories carousel is not AMP  or if the first publication does not use the scroll nav (some videos below, more to follow). 

Phone: Huawei P30 Pro, Android 10


https://user-images.githubusercontent.com/42241483/112207099-674e6d80-8c1f-11eb-9780-092971e18fbe.mp4




https://user-images.githubusercontent.com/42241483/112207090-64ec1380-8c1f-11eb-8a19-6f1f0cec9222.mp4

**Comment by mspr1nt** [comment]
https://user-images.githubusercontent.com/42241483/112207399-bbf1e880-8c1f-11eb-910a-6f575358816e.mp4

https://user-images.githubusercontent.com/42241483/112207662-12f7bd80-8c20-11eb-8b0d-28f40fa72c77.mp4

**Comment by mspr1nt** [comment]
Additional info: might be worth noting that recently I have noticed an inconsistency in the AMP bolt being displayed in the search results (for top stories and in general). This is present for pages that are definitely AMP and would otherwise show as AMP (with the bolt). The bolt is absent when using the search widgets when the "open web pages in-app" option is toggled off. 

Issue was noted before - most notably in September 2020 around 20/21 September when there were rogue mobile indexing issues.

**Comment by mspr1nt** [comment]
Finally, if it's helpful to compare versions/seek out changes, the notable dates for the trajectories of AMP results (as an average % of total) in the SERPS  (US specific in this case), as follows: 

2 March (decline begins)
4 March (drastic decrease)
8 March (moderate increase)
10 March (decline)
17 March (sharp decline)
20/21 March (moderate increase)

File: extensions/amp-viewer-integration/0.1/touch-handler.js (hunk 5 above)
Comment (rcebulko) [reply extensions/amp-viewer-integration/0.1/touch-handler.js:141]:
```suggestion
    if (e?.shouldViewerCancelPropagation) {
```

Comment (kristoferbaxter) [reply extensions/amp-viewer-integration/0.1/touch-handler.js:141]: Keep forgetting about this newer syntax. Is there a codemod we could apply on PRs to recommend updated syntax suggestions like this?

**Pull request merged**